`wrapper` and `asyncWrapper` options

overlookmotel opened the pull request:
PR for #4. Work in progress.

overlookmotel committed:
`wrapper` and `asyncWrapper` options

overlookmotel committed:
`asyncWrapCondition` option

danielstjules commented on index.js:
```diff
@@ -47,41 +53,95 @@ function wrapStaticMethods(klass, methodNames) {
 }
 
-function wrapFunctions(target, methodNames) {
+function wrapFunctions(target, options, isStatic) {
```
Looks like you switched to using the leading-underscore naming convention for pricate helpers, so this should probably be `_wrapFunctions` ?

overlookmotel replied:
I've pushed another commit that makes this change (and adds code comments on what function does).

danielstjules commented:
I think I mostly have stylistic comments regarding line length and such, but those are things I can change after merge. I might also make a few changes to the options validation.

Looks good! :)

overlookmotel committed:
Code style: underscore on `_wrapFunctions` name

overlookmotel committed:
`wrapCondition` option

overlookmotel commented:
@danielstjules Thanks for the comments. I've changed `wrapFunctions` to `_wrapFunctions` as suggested.

As far as code style is concerned, well I didn't really do the underscore style on purpose - just needed _something_ to call `_wrapStaticMethods()` and `_wrapInstanceMethods()`.

I've also added `options.wrapCondition` for filtering which generators to wrap for symmetry with `options.asyncWrapCondition`.

Are you happy with these condition options?

I'm getting really tight for time - project deadline looming - so wondered if you wanted to take over from here? If you're also too busy, no problem. But I'm unlikely to get time before Sept now to finish off with tests and the `.options()` method. It's rather crap of me to say "I've done the fun stuff, now you write the tests!" but I'm just being realistic. If you're in no hurry to get it merged, no problem - it can wait until I am freed up a bit.

Please let me know what you think, and thanks for entertaining this PR which complicates the API quite a bit.

danielstjules commented:
Sorry just picking this up now 😅 
The library probably isn't used anywhere, but I can pickup these final changes. Thanks for your help! I'll probably leave comments here and there as I review and make changes.

danielstjules commented:
I'm thinking wrapCondition/asyncWrapCondition are perhaps unnecessary since the user is already able to supply an array of methods to wrap. That array can be built dynamically, eliminating the need for these options.

danielstjules commented on index.js:
```diff
@@ -47,41 +53,117 @@ function wrapStaticMethods(klass, methodNames) {
-    if (target[key].constructor.name === 'GeneratorFunction') {
-      target[key] = Promise.coroutine(target[key]);
+    if (isGeneratorFunction) {
+      if (options.wrapper) target[key] = options.wrapper(target[key]);
```
Shouldn't the generator use the `asyncWrapper`, not `wrapper` ?
Edit: nvm, though I'm not sure I understand the naming convention used

danielstjules commented:
Started reviewing and working on a local branch at https://github.com/danielstjules/async-class/tree/options. With this current implementation, a user who wished to supply a different promise/coroutine implementation would still need to resort to some helper, otherwise it would be quite verbose:

``` javascript
let asyncClass = require('async-class');
let co = require('co');

function wrap(klass, methodNames) {
  return asyncClass.wrap(klass, methodNames, {
    asyncWrapper: ??,
    wrapper: co.wrap
  });
};
```

Unfortunately, I don't have too much time for this either, and I'm not as keen on the two optional arguments API. Luckily, a co-based fork already exists at https://github.com/viniciusps2/async-class-co :) And ES7 async/await is around the corner, so this module probably won't be useful for that much longer (not that anyone uses it, ha)

Appreciate the help with this! Hope you don't mind if I close this, and you're of course welcome to push a fork with this functionality!
